File: build_user/app/tmp/kapt3/stubs/debug/com/assistx/monitor/ui/history/HistoryFragment.java
package com.assistx.monitor.ui.history;

@kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000F\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0004\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000b\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0010\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0005\u0018\u00002\u00020\u0001B\u0005\u00a2\u0006\u0002\u0010\u0002J\b\u0010\u0016\u001a\u00020\u0017H\u0002J\b\u0010\u0018\u001a\u00020\u0017H\u0016J\u001a\u0010\u0019\u001a\u00020\u00172\u0006\u0010\u001a\u001a\u00020\u001b2\b\u0010\u001c\u001a\u0004\u0018\u00010\u001dH\u0016J\b\u0010\u001e\u001a\u00020\u0017H\u0002J\b\u0010\u001f\u001a\u00020\u0017H\u0002J\u0010\u0010 \u001a\u00020\u00172\u0006\u0010!\u001a\u00020\u000bH\u0002R\u0010\u0010\u0003\u001a\u0004\u0018\u00010\u0004X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0014\u0010\u0005\u001a\u00020\u00048BX\u0082\u0004\u00a2\u0006\u0006\u001a\u0004\b\u0006\u0010\u0007R\u000e\u0010\b\u001a\u00020\tX\u0082.\u00a2\u0006\u0002\n\u0000R\u0010\u0010\n\u001a\u0004\u0018\u00010\u000bX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010\f\u001a\u00020\rX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000e\u001a\u00020\rX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000f\u001a\u00020\rX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u001b\u0010\u0010\u001a\u00020\u00118BX\u0082\u0084\u0002\u00a2\u0006\f\n\u0004\b\u0014\u0010\u0015\u001a\u0004\b\u0012\u0010\u0013\u00a8\u0006\""}, d2 = {"Lcom/assistx/monitor/ui/history/HistoryFragment;", "Landroidx/fragment/app/Fragment;", "()V", "_binding", "Lcom/assistx/monitor/databinding/FragmentHistoryBinding;", "binding", "getBinding", "()Lcom/assistx/monitor/databinding/FragmentHistoryBinding;", "historyAdapter", "Lcom/assistx/monitor/ui/history/HistoryAdapter;", "lastSummary", "Lcom/assistx/monitor/ui/history/AnalyticsSummaryUi;", "showCpu", "", "showDisk", "showRam", "viewModel", "Lcom/assistx/monitor/ui/history/HistoryViewModel;", "getViewModel", "()Lcom/assistx/monitor/ui/history/HistoryViewModel;", "viewModel$delegate", "Lkotlin/Lazy;", "observeData", "", "onDestroyView", "onViewCreated", "view", "Landroid/view/View;", "savedInstanceState", "Landroid/os/Bundle;", "setupChart", "setupChartFilterChips", "updateChart", "summary", "app_debug"})
public final class HistoryFragment extends androidx.fragment.app.Fragment {
    @org.jetbrains.annotations.Nullable
    private com.assistx.monitor.databinding.FragmentHistoryBinding _binding;
    @org.jetbrains.annotations.NotNull
    private final kotlin.Lazy viewModel$delegate = null;
    private com.assistx.monitor.ui.history.HistoryAdapter historyAdapter;
    private boolean showCpu = true;
    private boolean showRam = true;
    private boolean showDisk = true;
    @org.jetbrains.annotations.Nullable
    private com.assistx.monitor.ui.history.AnalyticsSummaryUi lastSummary;
    
    public HistoryFragment() {
        super();
    }
    
    private final com.assistx.monitor.databinding.FragmentHistoryBinding getBinding() {
        return null;
    }
    
    private final com.assistx.monitor.ui.history.HistoryViewModel getViewModel() {
        return null;
    }
    
    @java.lang.Override
    public void onViewCreated(@org.jetbrains.annotations.NotNull
    android.view.View view, @org.jetbrains.annotations.Nullable
    android.os.Bundle savedInstanceState) {
    }
    
    /**
     * Setup chip filter CPU / RAM / DISK.
     * Jika tidak ada yang dipilih → tampilkan semua (fallback).
     */
    private final void setupChartFilterChips() {
    }
    
    private final void observeData() {
    }
    
    private final void setupChart() {
    }
    
    /**
     * Render grafik berdasarkan filter chip yang aktif.
     * Jika tidak ada yang dipilih → tampilkan semua dataset.
     */
    private final void updateChart(com.assistx.monitor.ui.history.AnalyticsSummaryUi summary) {
    }
    
    @java.lang.Override
    public void onDestroyView() {
    }
}

File: app/src/main/java/com/assistx/monitor/ui/history/HistoryFragment.kt
package com.assistx.monitor.ui.history

import android.graphics.Color
import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.assistx.monitor.R
import com.assistx.monitor.databinding.FragmentHistoryBinding
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class HistoryFragment : Fragment(R.layout.fragment_history) {

    private var _binding: FragmentHistoryBinding? = null
    private val binding get() = _binding!!
    private val viewModel: HistoryViewModel by viewModels()
    private lateinit var historyAdapter: HistoryAdapter

    // State filter grafik — default semua aktif
    private var showCpu = true
    private var showRam = true
    private var showDisk = true

    // Cache summary terakhir agar bisa re-render saat chip berubah
    private var lastSummary: AnalyticsSummaryUi? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentHistoryBinding.bind(view)

        historyAdapter = HistoryAdapter()
        binding.rvHistory.layoutManager = LinearLayoutManager(requireContext())
        binding.rvHistory.adapter = historyAdapter

        setupChart()
        setupChartFilterChips()

        binding.toggleGroup.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (!isChecked) return@addOnButtonCheckedListener
            val range = when (checkedId) {
                R.id.btnDaily -> "daily"
                R.id.btnWeekly -> "weekly"
                R.id.btnMonthly -> "monthly"
                else -> "daily"
            }
            viewModel.loadSummary(range)
        }

        observeData()
        viewModel.loadSummary("daily")
    }

    /**
     * Setup chip filter CPU / RAM / DISK.
     * Jika tidak ada yang dipilih → tampilkan semua (fallback).
     */
    private fun setupChartFilterChips() {
        val onChipChanged = {
            val cpuChecked = binding.chipCpu.isChecked
            val ramChecked = binding.chipRam.isChecked
            val diskChecked = binding.chipDisk.isChecked

            // Jika tidak ada yang dipilih, tampilkan semua
            val noneSelected = !cpuChecked && !ramChecked && !diskChecked
            showCpu = if (noneSelected) true else cpuChecked
            showRam = if (noneSelected) true else ramChecked
            showDisk = if (noneSelected) true else diskChecked

            lastSummary?.let { updateChart(it) }
        }

        binding.chipCpu.setOnCheckedChangeListener { _, _ -> onChipChanged() }
        binding.chipRam.setOnCheckedChangeListener { _, _ -> onChipChanged() }
        binding.chipDisk.setOnCheckedChangeListener { _, _ -> onChipChanged() }
    }

    private fun observeData() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.summary.collectLatest { summary ->
                if (_binding == null || summary == null) return@collectLatest

                lastSummary = summary

                // Update Average Usage Cards
                val globalAvgCpu = summary.avgCpus.filter { it > 0 }.average()
                    .let { if (it.isNaN()) 0.0 else it }
                val globalAvgMem = summary.avgMems.filter { it > 0 }.average()
                    .let { if (it.isNaN()) 0.0 else it }
                val globalAvgStorage = summary.avgStorages.filter { it > 0 }.average()
                    .let { if (it.isNaN()) 0.0 else it }

                binding.tvAvgCpu.text = String.format("%.1f%%", globalAvgCpu)
                binding.tvAvgRam.text = String.format("%.1f%%", globalAvgMem)
                binding.tvAvgStorage.text = String.format("%.1f%%", globalAvgStorage)

                // Update timeline list
                historyAdapter.submitList(
                    summary.labels.zip(summary.onlineCounts)
                        .mapIndexed { index, (label, online) ->
                            TimelineItem(
                                timestamp = label,
                                online = online,
                                offline = summary.offlineCounts.getOrElse(index) { 0 },
                                avgCpu = summary.avgCpus.getOrElse(index) { 0.0 },
                                avgMem = summary.avgMems.getOrElse(index) { 0.0 },
                                avgStorage = summary.avgStorages.getOrElse(index) { 0.0 }
                            )
                        }
                )

                updateChart(summary)
            }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.isLoading.collectLatest { loading ->
                if (_binding != null) {
                    binding.progressBar.visibility = if (loading) View.VISIBLE else View.GONE
                }
            }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.error.collectLatest { msg ->
                if (_binding != null) {
                    binding.tvError.text = msg
                    binding.tvError.visibility = if (msg != null) View.VISIBLE else View.GONE
                }
            }
        }
    }

    private fun setupChart() {
        binding.historyChart.apply {
            description.isEnabled = false
            setTouchEnabled(true)
            isDragEnabled = true
            setScaleEnabled(true)
            setPinchZoom(true)
            setDrawGridBackground(false)
            legend.isEnabled = true
            legend.textColor = Color.GRAY
            legend.textSize = 11f

            xAxis.apply {
                position = XAxis.XAxisPosition.BOTTOM
                textColor = Color.GRAY
                setDrawGridLines(false)
                setDrawLabels(false) // label terlalu padat, sembunyikan
            }

            axisLeft.apply {
                textColor = Color.GRAY
                setDrawGridLines(true)
                axisMinimum = 0f
                axisMaximum = 100f
            }

            axisRight.isEnabled = false
        }
    }

    /**
     * Render grafik berdasarkan filter chip yang aktif.
     * Jika tidak ada yang dipilih → tampilkan semua dataset.
     */
    private fun updateChart(summary: AnalyticsSummaryUi) {
        if (_binding == null) return

        val cpuColor  = ContextCompat.getColor(requireContext(), R.color.brand_blue)
        val ramColor  = ContextCompat.getColor(requireContext(), R.color.warning_orange)
        val diskColor = ContextCompat.getColor(requireContext(), R.color.storage_blue)

        val datasets = mutableListOf<LineDataSet>()

        if (showCpu) {
            val entries = summary.avgCpus.mapIndexed { i, v -> Entry(i.toFloat(), v.toFloat()) }
            datasets.add(LineDataSet(entries, "CPU").apply {
                color = cpuColor
                setCircleColor(cpuColor)
                lineWidth = 2f
                circleRadius = 2f
                setDrawValues(false)
                mode = LineDataSet.Mode.CUBIC_BEZIER
            })
        }

        if (showRam) {
            val entries = summary.avgMems.mapIndexed { i, v -> Entry(i.toFloat(), v.toFloat()) }
            datasets.add(LineDataSet(entries, "RAM").apply {
                color = ramColor
                setCircleColor(ramColor)
                lineWidth = 2f
                circleRadius = 2f
                setDrawValues(false)
                mode = LineDataSet.Mode.CUBIC_BEZIER
            })
        }

        if (showDisk) {
            val entries = summary.avgStorages.mapIndexed { i, v -> Entry(i.toFloat(), v.toFloat()) }
            datasets.add(LineDataSet(entries, "DISK").apply {
                color = diskColor
                setCircleColor(diskColor)
                lineWidth = 2f
                circleRadius = 2f
                setDrawValues(false)
                mode = LineDataSet.Mode.CUBIC_BEZIER
            })
        }

        if (datasets.isEmpty()) {
            binding.historyChart.clear()
            binding.historyChart.invalidate()
            return
        }

        binding.historyChart.data = LineData(datasets.map { it })
        binding.historyChart.invalidate()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}

data class TimelineItem(
    val timestamp: String,
    val online: Int,
    val offline: Int,
    val avgCpu: Double,
    val avgMem: Double,
    val avgStorage: Double
)
